### PR TITLE
Functions instead of macros

### PR DESCRIPTION
**Pull request by SylvainCorlay**

### CODE REVIEW

**Comment by wjakob** [comment]
(Not sure that this is a big improvement, it looks less readable)
cc @aldanor

**Comment by SylvainCorlay** [comment]
functions over macros: type checking, namespaces, const correctness, nicer to look at.

**Comment by jagerman** [comment]
As functions, the full `::pybind11::detail::whatever` name isn't needed: they can all be shortened to `detail::whatever`.

For that matter, I think the functions should live in the `pybind11::detail` namespace: these aren't meant to be callable externally anyway (at least I assume so, given the trailing underscore), and that's what `detail` is for.

**Comment by jagerman** [comment]
In support of the idea behind this PR, I'm also not a big fan of function-like macros where a small inline function does the job just as well.

**Comment by SylvainCorlay** [comment]
+1 on detail for the type. Will change this when I get close to a keyboard.

**Comment by aldanor** [comment]
The reason behind using macros was so it looks more like NumPy C API; inline functions are reasonable as well I guess, as long as they stay within `detail` namespace.

Note: it's array *descriptor* and not *description*. Might also use `array_descr_proxy` name for consistency with `array_proxy`.

File: include/pybind11/numpy.h [hunk 1]
```diff
@@ -396,65 +407,65 @@ class array : public buffer {
...
     /// Base object
     object base() const {
-        return reinterpret_borrow<object>(PyArray_GET_(m_ptr, base));
+        return reinterpret_borrow<object>(array_proxy(m_ptr)->base);
     }
 
     /// Dimensions of the array
-    const size_t* shape() const {
-        return reinterpret_cast<const size_t *>(PyArray_GET_(m_ptr, dimensions));
+    inline const size_t* shape() const {
```
Comment (aldanor) [inline include/pybind11/numpy.h]: This `inline` and a few more below are redundant, no?

**Comment by dean0x7d** [comment]
Adding `inline` to member functions is either redundant or compiler dependent. They're usually handled with compatibility macros (`STRONG_INLINE`, `ALWAYS_INLINE`), if that is even needed here.

**Comment by SylvainCorlay** [comment]
> Adding inline to member functions is either redundant or compiler dependent. They're usually handled with compatibility macros (STRONG_INLINE, ALWAYS_INLINE), if that's even needed here.

STRONG_INLINE and ALWAYS_INLINE are compiler-specific things. I generally dislike non-standard things. The `inline` keyword is the standard keyword showing intent of inlining. It is generally good practice for header-only library to specify non-template methods as inline.

File: include/pybind11/numpy.h (hunk 1 above)
Comment (SylvainCorlay) [inline include/pybind11/numpy.h]: see comment in the PR discussion.

**Comment by SylvainCorlay** [comment]
> Note: it's array descriptor and not description. Might also use array_descr_proxy name for consistency with array_proxy.

Changing  for `array_description_proxy`.

**Comment by aldanor** [comment]
> Changing for `array_description_proxy`.

`array_descriptor_proxy`

**Comment by SylvainCorlay** [comment]
> > Changing for array_description_proxy.
> array_descriptor_proxy

That is what I meant :)

**Comment by jagerman** [comment]
`inline` is implied by the standard for member function defined entirely within a class definition like these--just like it's implied for template functions.  It seems inconsistent to omit one but not the other.

**Comment by dean0x7d** [comment]
Member functions defined completely inside a class definition are implicitly `inline`, so by the letter of the standard there is no benefit to making them explicitly `inline`.

**Comment by aldanor** [comment]
@jagerman @dean0x7d Ditto re: `inline`.

**Comment by SylvainCorlay** [comment]
> Note: it's array descriptor and not description. Might also use array_descr_proxy name for consistency with array_proxy.

So this is ready for review. I kept the new inline function in the general pybind namespace (which is already more scoped than a macro). The reason is that I intend to use them directly in xtensor when pybind 2 is released.

**Comment by jagerman** [comment]
> I kept the new inline function in the general pybind namespace (which is already more scoped than a macro). The reason is that I intend to use them directly in xtensor when pybind 2 is released.

This is a little bit worrying: these are then public `pybind11` functions that return a `pybind11::detail` pointer, but that seems wrong: the only way you can use them externally is via external references to internal data structures.  Is there some other way to expose the same functionality without needing to expose an internal data structure?

**Comment by SylvainCorlay** [comment]
Less than before anyway since the macros were exposing the same thing and were not scoped at all. Besides `detail` is not hidden either...

**Comment by wjakob** [comment]
This PR creates new non-documented public API functions that expose internal data structures from the ``detail`` namespace. Since the main reason for this seems to be for the benefit of the ``xtensor`` library, I would suggest that ``xtensor`` just plugs directly into the ``detail`` namespace and private macros. For instance, there is no reasons why (with a suitable ``using`` declaration) it can't use the current macros that are IMHO also more readable.

**Comment by SylvainCorlay** [comment]
Alright, I would be ready to document things. Will move this to detail.
I was pointing out that it was also the case with the api I am replacing.

**Comment by jagerman** [comment]
I don't like the current macros: they aren't prefixed with PYBIND11 and have somewhat generic names (as @aldanor said, mirroring the numpy interface).  There seems some real potential for name clashes there.

**Comment by SylvainCorlay** [comment]
All is in detail. Ready to go.

**Comment by aldanor** [comment]
Looks good to me (the extra added verbosity aside).

**Comment by SylvainCorlay** [comment]
Merge?

**Pull request merged**